### PR TITLE
Make consistent behavior on zeros equality on floating point types

### LINKED ISSUE #3509: Make consistent behavior on zeros equality on floating point types
**Describe the bug**

For floating point types, we treat positive and negative zeros as same on `build_compare`, but different on comparison kernels which use `total_cmp`. It's better to be consistent the behavior with comparison kernels.

**To Reproduce**

**Expected behavior**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #3509.

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by tustvold** [comment]
I'm not sure about this, as it means we no longer are comparing with respect to a standard predicate but one of our own devising. Why special case zero, and not other values like NaNs?

I'd also be interested to know what impact this has on benchmarks.

FWIW If we do make this change, we will need to make changes to normalise within the row format, along with potentially in other places also. Nothing insurmountable, just noting it

File: arrow-ord/src/ord.rs [hunk 1]
```diff
@@ -21,32 +21,21 @@ use arrow_array::types::*;
...
     right: &dyn Array,
 ) -> DynComparator
 where
-    T::Native: Ord,
+    T::Native: ArrowNativeTypeOp,
 {
     let left: PrimitiveArray<T> = PrimitiveArray::from(left.data().clone());
     let right: PrimitiveArray<T> = PrimitiveArray::from(right.data().clone());
-    Box::new(move |i, j| left.value(i).cmp(&right.value(j)))
+    Box::new(move |i, j| left.value(i).compare(right.value(j)))
```
Comment (tustvold) [inline arrow-ord/src/ord.rs:38]: 👍 regardless this is a good change

**Comment by viirya** [comment]
NaNs are treated as equal by total ordering. I guess total ordering needs to give a comprehensive ordering for possible floating point values. But in practice computation, we don't actually separate positive and negative zeros.

**Comment by tustvold** [comment]
> NaNs are treated as equal by total ordering

Not the ordering we use currently, they're ordered based on their constituent bits, NaNs with different byte representations will not compare equal

**Comment by viirya** [comment]
> Not the ordering we use currently, they're ordered based on their constituent bits, NaNs with different byte representations will not compare equal

We have NaN equality test to verify that they are equal. I also did a quick verification in rust playground:

```rust
fn main() {
    let a = f32::NAN;
    let b = f32::NAN;
    
    println!("a == b: {}", a.to_bits() == b.to_bits());
}
```

Output:
```
a == b: true
```

**Comment by tustvold** [comment]
f32::NaN always returns the same NaN bytes, if you get a NaN by other means such that they have different bit representations you will see the difference

**Comment by viirya** [comment]
I see. That explains why these NaNs are equal. I roughly remember that from JVM experience NaN values' bits are different so I was a bit surprised to see they are equal in above test/play-ground. If there are other bit patterns in Rust that will be seen as NaN too, then it is not guaranteed to be equal.

NaNs should be treated as equal in computation too, like zeros.

Either adding NaN-specific condition like zero, or we avoid such things here and require users to handle it before calling arrow kernels. For example, replacing negative zeros with positive zeros, normalizing NaNs with standard f32::NaN (f64, f16 too).

File: arrow-array/src/arithmetic.rs [hunk 2]
```diff
@@ -363,15 +363,25 @@ macro_rules! native_type_float_op {
...
             #[inline]
             fn compare(self, rhs: Self) -> Ordering {
-                <$t>::total_cmp(&self, &rhs)
+                if self.abs() == $zero && rhs.abs() == $zero {
+                    // `total_cmp` treats positive zero and negative zero as different.
+                    // But for computation system, it usually treats them as equal.
+                    Ordering::Equal
+                } else {
+                    <$t>::total_cmp(&self, &rhs)
+                }
```
Comment (viirya) [inline arrow-array/src/arithmetic.rs]: I removed these changes.

File: arrow-ord/src/comparison.rs [hunk 3]
```diff
@@ -685,6 +691,8 @@ where
 /// If `simd` feature flag is not enabled:
 /// For floating values like f32 and f64, this comparison produces an ordering in accordance to
 /// the totalOrder predicate as defined in the IEEE 754 (2008 revision) floating point standard.
+/// Note that totalOrder treats positive and negative zeros are different. If it is necessary
+/// to treat them as equal, please normalize zeros before calling this kernel.
```
Comment (viirya) [inline arrow-ord/src/comparison.rs:695]: I updated these docs to make the behavior clear to users.

File: arrow-ord/src/ord.rs [hunk 4]
```diff
@@ -380,8 +358,8 @@ pub mod tests {
 
         let cmp = build_compare(&array, &array).unwrap();
 
-        assert_eq!(Ordering::Equal, (cmp)(0, 1));
-        assert_eq!(Ordering::Equal, (cmp)(1, 0));
+        assert_eq!(Ordering::Less, (cmp)(0, 1));
+        assert_eq!(Ordering::Greater, (cmp)(1, 0));
```
Comment (viirya) [inline arrow-ord/src/ord.rs:362]: `build_compare`'s behavior on zeros comparison is inconsistent with comparison kernels. Changed it to consistent.

**Review by tustvold — APPROVED** [review]
Thank you

**Comment by jhorstmann** [comment]
Looks good to me!

Just a note that the min/max aggregation kernels also use a different definition, I think following the postgres behavior of considering NaN to be greater than any other value.

**Comment by tustvold** [comment]
> I think following the postgres behavior of considering NaN to be greater than any other value

Yeah it is honestly baffling to me that they took so long to define a total ordering predicate, we now have a standard but few people follow it :sweat_smile:

**Pull request merged**